### DISCUSSION TITLE
HTTP/1, HTTP/2, HTTP/3 기반 실시간 통신의 성능 비교 및 최적화

### DISCUSSION THREAD

**Discussion by Namgyu11** [100-hours-a-week/discussions#38] [Deepdive]
### 목차

1. **프로토콜의 개념**
   - [1.1 프로토콜이란 무엇인가?](#11-프로토콜이란-무엇인가)
     - 데이터 통신에서 프로토콜의 정의와 필요성
   - [1.2 프로토콜의 기본 요소](#12-프로토콜의-기본-요소)
     - 데이터 전송, 에러 처리, 동기화 등 주요 기능
   - [1.3 네트워크 계층과 프로토콜](#13-네트워크-계층과-프로토콜)
     - OSI 7계층 모델과 각 계층에서의 프로토콜 역할

2. **HTTP의 기본 개념**
   - [2.1 HTTP란 무엇인가?](#21-http란-무엇인가)
     - 웹 상에서의 통신 프로토콜로서의 역할
   - [2.2 HTTP의 비상태성 및 기본 작동 원리](#22-http의-비상태성-및-기본-작동-원리)
     - 요청-응답 모델과 비상태성 설명
   - [2.3 TCP와 HTTP의 관계](#23-tcp와-http의-관계)
     - TCP/IP 모델에서 HTTP의 위치 및 의존성

3. **HTTP 버전별 진화 과정**
   - [3.1 HTTP/0.9: 최초의 HTTP](#31-http09-최초의-http)
     - 간단한 GET 요청과 응답의 구조
   - [3.2 HTTP/1.0: 현대 웹의 기초](#32-http10-현대-웹의-기초)
     - 상태 코드, 메서드 확장, 헤더 도입
   - [3.3 HTTP/1.1: 가장 널리 사용된 버전](#33-http11-가장-널리-사용된-버전)
     - 지속 연결, 파이프라이닝, 캐싱 등의 개선점
   - [3.4 HTTP/2: 멀티플렉싱과 성능 최적화](#34-http2-멀티플렉싱과-성능-최적화)
     - 바이너리 프로토콜, 헤더 압축, 서버 푸시
   - [3.5 HTTP/3: QUIC 기반의 새로운 HTTP](#35-http3-quic-기반의-새로운-http)
     - UDP 기반 QUIC, 0-RTT, 보안과 성능 향상

4. **HTTP 버전별 차이점 분석**
   - [4.1 연결 관리](#41-연결-관리)
     - 각 버전에서의 연결 방식 차이 (새로운 연결 vs 지속 연결 vs QUIC)
   - [4.2 성능 최적화](#42-성능-최적화)
     - 멀티플렉싱, 헤더 압축, 서버 푸시 등 성능 관련 기능 비교
   - [4.3 보안](#43-보안)
     - 각 버전에서의 보안 강화 (HTTPS, TLS 통합 등)
   - [4.4 데이터 전송 방식](#44-데이터-전송-방식)
     - 텍스트 vs 바이너리 vs UDP 기반 전송 비교
   - [4.5 실제 적용 사례](#45-실제-적용-사례)
     - 각 버전이 현재의 웹 환경에서 어떻게 적용되는지 예시

5. **HTTP 버전 선택 기준**
   - [5.1 프로젝트 특성에 따른 HTTP 버전 선택 가이드](#51-프로젝트-특성에-따른-http-버전-선택-가이드)
     - 웹사이트 성능 최적화, 실시간 통신, 모바일 환경 등 고려 사항
   - [5.2 HTTP/2와 HTTP/3 적용 사례](#52-http2와-http3-적용-사례)
     - 실시간 데이터 처리, 빠른 페이지 로딩을 위한 선택 기준
   - [5.3 전환의 어려움과 대응 방안](#53-전환의-어려움과-대응-방안)
     - HTTP/1.x에서 HTTP/2, HTTP/3로 전환할 때의 고려 사항

6. **결론**
   - [6.1 HTTP의 진화와 웹 기술의 발전](#61-http의-진화와-웹-기술의-발전)
   - [6.2 프로젝트에 맞는 HTTP 버전 선택의 중요성](#62-프로젝트에-맞는-http-버전-선택의-중요성)
   - [6.3 향후 HTTP의 발전 방향 전망](#63-향후-http의-발전-방향-전망)

---

# 1. 프로토콜

### a. 기본적인 정의

- **프로토콜**은 컴퓨터 네트워크에서 서로 다른 시스템 간의 통신을 가능하게 하는 규칙과 절차의 집합입니다. 
  - 데이터를 한 시스템에서 다른 시스템으로 전달할 때 **어떻게 보내고, 어떻게 받으며, 이를 처리하는 방법**을 정의한 규칙입니다.

### b. 일상적 예시로 이해하기

- 프로토콜을 쉽게 이해하기 위해, 사람 간의 일상적인 대화를 예로 들어봅시다.
  - 만약 두 사람이 서로 다른 언어를 사용한다면, 의사소통이 불가능합니다. 원활한 대화를 위해서는 공통으로 이해할 수 있는 **언어(문법과 어휘)**를 사용해야 합니다.
  - 또한 대화의 순서와 매너도 중요합니다. 한 사람이 말을 할 때, 다른 사람은 끝까지 듣고 대답해야 합니다.
  - 네트워크 상의 통신에서도 마찬가지로, 컴퓨터들이 서로 데이터를 주고받기 위해 동일한 언어(프로토콜)를 사용하며, **순서와 절차**를 지켜야 합니다. 이 언어와 절차가 바로 프로토콜입니다.

### c. 프로토콜의 필요성

- 네트워크 상에서 다양한 시스템들이 서로 데이터를 주고받을 때 각 시스템이 자신만의 방식으로 데이터를 처리하면 통신이 불가능합니다. 이 때문에 프로토콜을 사용해야 합니다. 그 필요성을 다음과 같이 설명할 수 있습니다:

1. **호환성**
   - 서로 다른 장치나 운영 체제, 네트워크 기술을 사용하는 컴퓨터들이 데이터를 주고받기 위해서는 **공통된 규칙**이 필요합니다. 이 규칙을 정의한 것이 프로토콜입니다. 예를 들어, Windows PC와 Mac 간 이메일을 주고받거나, Android와 iOS 스마트폰이 같은 웹사이트에 접속할 수 있는 것은 **공통된 프로토콜** 덕분입니다.

2. **데이터의 무결성과 신뢰성 보장**
   - 통신 중 데이터가 손실되거나 손상되지 않도록 **재전송**이나 **오류 수정** 방법을 규정하여 신뢰성을 보장합니다. 예를 들어, **TCP(Transmission Control Protocol)**는 데이터를 패킷 단위로 쪼개 전송하며, 각 패킷이 정확히 도착했는지 확인하는 메커니즘을 갖추고 있습니다.

3. **통신 속도의 최적화**
   - 데이터를 주고받는 과정에서 **속도**는 중요한 요소입니다. 프로토콜은 데이터를 한 번에 보낼 수 있는 크기나 전송 속도를 조정하여 **네트워크 자원을 효율적으로 사용**하고 최적의 속도로 데이터를 전송합니다.

4. **동기화와 통신 흐름 관리**
   - 프로토콜은 데이터를 주고받는 **순서와 타이밍**을 정의하여 통신 흐름을 관리합니다. 이를 통해 데이터가 중복되거나 손실되지 않도록 통신을 동기화할 수 있습니다.

5. **보안**
   - 프로토콜은 데이터를 **암호화**하고 이를 **복호화**하는 방식을 규정하여, 외부의 침입이나 도청으로부터 데이터를 보호할 수 있도록 설계됩니다. 예를 들어, HTTPS는 HTTP에 SSL/TLS 암호화를 더해 웹에서의 보안 통신을 보장합니다.

---

## 1.2 프로토콜의 기본 요소

> 프로토콜의 기본 요소는 데이터가 네트워크를 통해 원활하게 전송되도록 하는 규칙과 메커니즘을 정의합니다. 각 요소는 전송의 신뢰성, 보안성, 그리고 효율성을 보장하기 위한 핵심 기능입니다.

### 1. 데이터 전송

- **TCP**(RFC 793)에 따르면, 데이터 전송은 다음과 같은 절차를 통해 관리됩니다.

#### a. 데이터 분할 (Segmentation)
- **TCP**는 데이터를 전송하기 전에 **세그먼트(segment)**로 분할하여 전송합니다. 네트워크 상에서 대용량 데이터를 작은 패킷으로 나누어 보내는 방식으로 전송 속도와 효율성을 높입니다. 수신 측에서는 이 세그먼트를 다시 **재조립(reassembly)**하여 원래 데이터를 복원합니다.

#### b. 전송 방식: 연결 지향과 비연결 지향
- **연결 지향(Connection-oriented)**: TCP는 데이터를 전송하기 전에 **3-way handshake**로 송신자와 수신자가 연결을 설정한 후 데이터를 주고받습니다. 이를 통해 신뢰성 있는 데이터 전송이 보장됩니다.
- **비연결 지향(Connectionless)**: **UDP**(RFC 768)는 별도의 연결 설정 없이 패킷을 바로 보냅니다. 이는 속도는 빠르지만 신뢰성은 낮습니다.

#### c. 데이터 재조립
- 수신자는 각 패킷에 부여된 **시퀀스 번호**를 바탕으로 데이터를 올바른 순서로 재조립합니다. **TCP**는 패킷이 순서대로 도착하지 않거나 손실되었을 경우 **재전송** 요청을 통해 데이터의 무결성을 보장합니다.

### 2. 에러 처리

> 에러 처리는 전송 중 데이터가 손상되거나 손실되었을 때 이를 감지하고 복구하는 메커니즘입니다.

#### a. 에러 검출 (Error Detection)
- **체크섬(Checksum)**: **TCP**는 송신자가 패킷을 전송할 때 **체크섬 필드**를 사용하여 데이터의 무결성을 검증합니다. 수신자는 체크섬 값을 통해 데이터가 손상되었는지 여부를 판단할 수 있습니다.

#### b. 에러 복구 (Error Recovery)
- **재전송**: 수신자가 손상된 패킷을 감지하면 송신자에게 **ACK(Acknowledgement)** 메시지를 보내지 않거나 **NAK(Negative Acknowledgement)** 메시지를 보내 패킷의 재전송을 요청합니다.

### 3. 동기화

> 동기화는 송신자와 수신자 간에 데이터가 순서대로 전달되고 올바르게 해석되도록 조율하는 과정입니다. 특히 연결 지향 프로토콜에서는 연결 설정, 유지, 종료 과정에서 동기화가 중요합니다.

#### a. 연결 설정 (3-Way Handshake)
- **TCP**는 데이터를 주고받기 전에 **3-way handshake** 과정을 통해 연결을 설정합니다. 송신자가 **SYN** 패킷을 보내 연결을 요청하면, 수신자는 **SYN-ACK** 패킷으로 응답하고, 송신자가 **ACK** 패킷을 보내 연결이 완료됩니다.

#### b. 데이터 전송 동기화
- 데이터 전송 중 **시퀀스 번호**를 사용하여 패킷의 순서대로 도착 여부를 확인하며, 손실된 패킷은 **재전송**됩니다.

### 4. 흐름 제어와 혼잡 제어

> 데이터 전송 속도를 조절하여 네트워크 자원을 효율적으로 사용하고, 네트워크 혼잡을 방지하는 기능입니다. 이는 데이터 손실과 네트워크 과부하를 방지하는 중요한 역할을 합니다.

#### a. 흐름 제어 (Flow Control)
- **TCP**는 **슬라이딩 윈도우(Sliding Window)** 기법을 사용하여 송신자가 수신자의 처리 능력에 맞게 데이터 전송 속도를 조절합니다. 수신자는 처리 가능한 **윈도우 크기**를 송신자에게 알리고, 송신자는 이에 맞춰 데이터를 보냅니다.

#### b. 혼잡 제어 (Congestion Control)
- **TCP**는 네트워크 트래픽이 많아져 혼잡이 발생하면 전송 속도를 조절합니다. **AIMD(Additive Increase, Multiplicative Decrease)** 알고리즘을 사용해 네트워크 상태에 따라 속도를 줄이거나 늘립니다.

### 5. 보안 (Security)

> 프로토콜은 네트워크에서 데이터의 기밀성과 무결성을 보장하기 위한 보안 기능을 포함합니다. **TLS(Transport Layer Security)** 프로토콜을 사용해 데이터 전송 중 암호화를 수행하며, 인증과 무결성 검사를 통해 보안을 강화합니다.

#### a. 암호화 (Encryption)
- **TLS 1.3**은 전송되는 데이터를 암호화하여 외부에서 데이터가 도청되거나 변조되는 것을 방지합니다. **HTTPS**는 **TLS**를 사용해 웹 브라우저와 서버 간의 보안 통신을 제공합니다.

#### b. 무결성 (Integrity)
- **해시 함수**와 **MAC(Message Authentication Code)**를 사용하여 데이터가 중간에서 변경되지 않았음을 보장합니다.

#### c. 인증 (Authentication)
- **SSL/TLS 인증서** 시스템을 통해 서버와 클라이언트가 서로 신뢰할 수 있음을 인증합니다. 인증서는 제3자인 **CA(Certificate Authority)**가 발급합니다.

**Comment by startupcodekr** [comment]
왜 이걸 해야하는지 프로젝트 문제점 시나리오 기반으로 제목을 작성해주세요!
